1.21

SettingDust opened the pull request.

SettingDust committed:
chore(deps): bump dependencies

SettingDust committed:
chore(build): bump gradle and loom

SettingDust committed:
chore: add parchmentmc for better param naming

SettingDust committed:
chore: disable jei, rei, and kubejs compat for now

SettingDust committed:
feat(contextual): ported contextual with codec

SettingDust committed:
fix(contextual): add constant check to min, max

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(contextual): add `ContextualConditionsHolder` for reducing records' duplicate code

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: rewrite part of `ItemStackHolder`, `PostAction`, `LycheeRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: rewrite part of context

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: complete json, generic, action contexts

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: try adding key based value map codec for context

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: add default value for random

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor(contextual): switch to new context

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: using common holders as abstract classes

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipes) add serializer and type for `AnvilCraftingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: remove catalogs for simple class and specify the default namespace to `lychee`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: port part of BlockKKeyRecipe

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: correct the generic type

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: port part of BlockKeyRecipe

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee commented on src/main/java/snownee/lychee/recipes/AnvilCraftingRecipe.java:
```diff
@@ -0,0 +1,199 @@
+								return DataResult.error(() -> "level_cost must be greater than 0");
+							}
+							return DataResult.success(it);
+						}).orElse(1).forGetter(AnvilCraftingRecipe::levelCost),
```
`ExtraCodecs.validate`

Snownee commented on gradle.properties:
```diff
@@ -3,23 +3,24 @@ org.gradle.daemon=false
+modmenu_version=9.0.0-pre.1
 architectury_version=9.1.12
-kiwi_version=6.0.0
+kiwi_version=13.0.0
```
Kiwi该更新了。另外1.20.4+将强制依赖Kiwi

Snownee commented on src/main/java/snownee/lychee/context/recipes/AnvilContext.java:
```diff
@@ -0,0 +1,61 @@
+	/**
+	 * Both inputs in anvil
+	 */
+	private final Pair<ItemStack, ItemStack> input;
```
应该是ItemContext的一部分，不明白为什么放在这里

SettingDust replied:
https://github.com/Snownee/Lychee/blob/1.20-fabric/src/main/java/snownee/lychee/anvil_crafting/AnvilContext.java#L13-L14

因为他俩在这里

Snownee commented on src/main/resources/assets/lychee/lang/en_us.json:
```diff
@@ -71,9 +71,9 @@
-  "result.lychee.pass": "§7- ",
+  "result.lychee.true": "§a✔ ",
+  "result.lychee.false": "§4✕ ",
+  "result.lychee.default": "§7- ",
```
修改键名时请将其他语言一并修改

Snownee commented on src/main/resources/lychee.mixins.json:
```diff
@@ -1,58 +1,60 @@
+        "LightningBoltMixin",
+        "LightPredicateAccess",
+        "LocationCheckMixin",
+        "LocationPredicate$PositionPredicateAccessor",
```
使用Access作为结尾。明明尽量简短

Snownee commented on src/main/resources/lychee.mixins.json:
```diff
@@ -1,58 +1,60 @@
+        "fabric.ResourceManagerHelperImplMixin",
+        "fabric.ShapedRecipeMixin",
+        "predicates.LocationCheckAccessor",
+        "predicates.LocationPredicate$PositionPredicateAccessor",
```
使用Access作为结尾。明明尽量简短

Snownee commented on src/main/java/snownee/lychee/recipes/anvil_crafting/AnvilCraftingRecipe.java:
```diff
@@ -25,23 +29,26 @@
 import snownee.lychee.util.json.JsonPointer;
 
-public class AnvilCraftingRecipe extends LycheeRecipe<AnvilContext> implements Comparable<AnvilCraftingRecipe> {
+public record AnvilCraftingRecipe implements LycheeRecipe<AnvilContext>, Comparable<AnvilCraftingRecipe> {
```
原先通过继承改写的匹配逻辑如今要怎么实现？（EverPotion）

SettingDust replied:
没理解XD

Snownee commented on src/main/java/snownee/lychee/recipes/AnvilCraftingRecipe.java:
```diff
@@ -0,0 +1,199 @@
+import snownee.lychee.util.recipe.LycheeRecipeCommonHolder;
+import snownee.lychee.util.recipe.LycheeRecipeSerializer;
+
+public record AnvilCraftingRecipe(
```
原先通过继承改写的匹配逻辑如今要怎么实现？（EverPotion）

SettingDust replied:
没理解XD

Snownee replied:
看这个实现：https://github.com/Snownee/EverPotion/blob/1.19.2-fabric/src/main/java/snownee/everpotion/crafting/FallbackCoreRecipe.java

SettingDust replied:
手写？

Snownee replied:
显然不行。另外`RecordCodecBuilder`并不是只能接受record

SettingDust replied:
> 显然不行。

那我倾向于这个上下文使用一个 `AnvilContext`，而不是继承
EDIT: 可能会把 AnvilContext 命名成更抽象的名称

> 另外`RecordCodecBuilder`并不是只能接受record

知道的

SettingDust replied:
啊，看错了，昨天看成 Context 了，那就转换成 class 呗

Snownee commented on src/main/java/snownee/lychee/util/recipe/LycheeRecipeType.java:
```diff
@@ -0,0 +1,101 @@
+	public boolean canPreventConsumeInputs;
+	public boolean hasStandaloneCategory = true;
+
+	private Supplier<Boolean> isEmpty = Suppliers.memoize(() -> recipes.isEmpty());
```
为什么要懒加载？应改名empty或emptyState

SettingDust replied:
没什么实际理由，因为原先那里就是刷新时全部加载一遍，我感觉没必要，就这样了

Snownee replied:
但之前必定已经执行过`refreshCache`，所有配方已经完全加载了

Snownee replied:
仍未改名

Snownee commented on src/main/java/snownee/lychee/util/context/LycheeContext.java:
```diff
@@ -0,0 +1,51 @@
+import snownee.lychee.util.SerializableType;
+
+@SuppressWarnings("unchecked")
+public class LycheeContext extends HashMap<LycheeContextType<?>, LycheeContextValue<?>> implements DummyContainer {
```
继承HashMap的必要性？而不是作为成员？在我看来，只需要get和put方法。如必要，为什么不用IdentityHashMap

SettingDust replied:
没有必要
为什么要用 IdentityHashMap？

Snownee replied:
速度更快

Snownee commented on src/main/java/snownee/lychee/util/context/LycheeContext.java:
```diff
@@ -0,0 +1,51 @@
+
+@SuppressWarnings("unchecked")
+public class LycheeContext extends HashMap<LycheeContextType<?>, LycheeContextValue<?>> implements DummyContainer {
+	public static final Codec<LycheeContext> CODEC =
+			new KeyDispatchedMapCodec<LycheeContextType<?>, LycheeContextValue<?>>(
+					LycheeRegistries.CONTEXT.byNameCodec(),
+					it -> DataResult.success(it.type()),
+					it -> it instanceof SerializableType<?> serializableType
+						  ? DataResult.success((Codec<? extends LycheeContextValue<?>>) serializableType.codec())
+						  : DataResult.error(() -> it + " isn't serializable"),
+					LycheeRegistries.CONTEXT
+			).codec().xmap(it -> {
+				final var context = new LycheeContext();
+				context.putAll(it);
+				return context;
+			}, Function.identity());
```
序列化是必要且可行的吗？ItemStackHolderCollection有瞬时性，世界中的ItemStackHolderCollection会随时变动/失效。我看不出把context分为不同组件的必要性。

SettingDust replied:
这样做的目的是需要存储到 marker 的数据不需要特别对待，只要实现接口会自动在 maker 里存储，而不是之前写特有逻辑

ItemContext 就没有实现序列化的接口，也就不会被存储

目前和原来是一致的，只保存 JsonContext 和 ActionContext。

Snownee replied:
问题是你说的实现序列化接口的Type在每种配方类型里都是必要的

SettingDust replied:
> 问题是你说的实现序列化接口的Type在每种配方类型里都是必要的

不是的，上下文类型里不是必须的

Snownee replied:
请举出反例

SettingDust replied:
> 请举出反例

除了 JsonContext 和 ActionContext 之外的 Context Type 都没有序列化的部分

Snownee replied:
但是有哪个LycheeContext是不需要ActionContext的？

SettingDust replied:
> 但是有哪个LycheeContext是不需要ActionContext的？

没理解，有必要的 Context 也有不一定存在的 Context

Snownee replied:
每个LycheeContext都有Level，都有Random，都有Loot Params，那你多加一步Map#get的目的是什么呢？每次创建LycheeContext时都要初始化一个HashMap，也是不小的开销

SettingDust replied:
不总是都有，比如 Codec 反序列化回来的时候就没有，需要调用点之后再填入这些上下文
至于开销，我不认为会造成可见的 CPU 时间占用，如果有性能问题再优化

Snownee replied:
1. 把LycheeContext反序列化回来的目的是执行Action，执行Action是一定需要Level, Random的
2. Lychee配方匹配是相当性能敏感的，用户可能每秒都在查询配方，一次爆炸/随机刻tick里面，如果有相应配方，被炸毁/tick的每一个方块都要创建一个新的LycheeContext，希望你能意识到这一点

SettingDust replied:
1. 是一定需要，但是反序列化之后，在调用点填入之前，这个上下文确实是缺失的
2. Sponge 的上下文是类似结构，https://github.com/SpongePowered/Sponge/blob/api-10/src/main/java/org/spongepowered/common/event/tracking/SpongeCauseStackFrame.java#L42-L43 ，且应用场景类似（事件上下文），但是引入了对象池 https://github.com/SpongePowered/Sponge/blob/api-10/src/main/java/org/spongepowered/common/event/tracking/PhaseTracker.java#L208-L210 ，所以如果有性能问题，那我会使用对象池

Snownee replied:
不只如此，每一次调用get()都是额外消耗。算了，之后我自己改吧

SettingDust replied:
> 不只如此，每一次调用get()都是额外消耗。算了，之后我自己改吧

这对性能的影响有多大尚未可知吧？但是这样的结构比起纯继承 + builder 耦合度更低，也更易扩展，

Snownee commented on src/main/java/snownee/lychee/util/predicates/BlockPredicate.java:
```diff
@@ -0,0 +1,160 @@
+package snownee.lychee.util.predicates;
```
从原版复制出来的原因？如有请注明

Snownee replied:
我认为理由不足，应该将增加的代码放入`BlockPredicateExtensions`

Snownee commented on src/main/java/snownee/lychee/util/recipe/LycheeRecipeSerializer.java:
```diff
@@ -0,0 +1,93 @@
+			Codec.list(PostAction.CODEC)
+				 .fieldOf("post")
+				 .orElse(Lists.newArrayList())
+				 .forGetter(LycheeRecipe::postActions);
```
1. 丧失了将JsonObject解析为单个元素的List的能力
2. 应使用`orElseGet`

Snownee commented on src/main/java/snownee/lychee/util/action/PostActionType.java:
```diff
@@ -0,0 +1,32 @@
+	RecordCodecBuilder<? extends PostAction<?>, ContextualCommonHolder> CONTEXTUAL_CODEC =
+			ContextualCommonHolder.CODEC
+					.fieldOf("contextual")
+					.orElse(new ContextualCommonHolder())
```
应使用`orElseGet`

Snownee commented on src/main/java/snownee/lychee/util/input/ItemStackHolderCollection.java:
```diff
@@ -0,0 +1,131 @@
+		for (final var holder : this) {
+			if (holder.getIgnoreConsumption() || holder.get().isEmpty()) continue;
+			final var count = holder.get().getCount();
+			holder.split(times);
```
为什么变成`split`了？

SettingDust replied:
`shrink` 是减少当前 ItemStack 的数量，`split` 也一样减少了，只是会返回被拿走的物品，可以拿到实际消耗的数量

Snownee replied:
`ItemStackHolder.split`和`ItemStackHolderCollection.split`行为不同，应该起不同的名字

SettingDust replied:
> `ItemStackHolder.split`和`ItemStackHolderCollection.split`行为不同，应该起不同的名字

我觉得 `ItemStackHolder.split` 和 `ItemStackHolderCollection.split` 是一个逻辑，都是减去数量之后返回分割出的 ItemStack，没明白命名需要怎么区分

Snownee replied:
`ItemStackHolderCollection.split`并不会减少`ItemStackHolderCollection`中总共的物品数量。`ItemStackHolder.split`应改为`shrink`

SettingDust replied:
> `ItemStackHolderCollection.split`并不会减少`ItemStackHolderCollection`中总共的物品数量。`ItemStackHolder.split`应改为`shrink`

你的意思是 `stacksNeedHandle` 也包含在里面所以不减少总共的物品数量吗？
我是觉得放到待处理区域已经和其他的不是一个状态了，也不属于任一 Holder 了

Snownee replied:
但语义是不同的

Snownee commented on src/main/java/snownee/lychee/util/recipe/LycheeRecipeSerializer.java:
```diff
@@ -0,0 +1,93 @@
+	Ingredient EMPTY_INGREDIENT = Ingredient.of(ItemStack.EMPTY);
+
+	RecordCodecBuilder<LycheeRecipe<?>, Boolean> HIDE_IN_VIEWER_CODEC =
+			Codec.BOOL.fieldOf("hide_in_viewer").orElse(false).forGetter(LycheeRecipe::hideInRecipeViewer);
```
`optionalFieldOf`？下面还有几个类似问题

SettingDust replied:
这里为什么需要 `optionalFieldOf`，没理解，它的类型也不是 `Optional`

Snownee replied:
使用`.optionalFieldOf("hide_in_viewer", false)`更直观，且序列化回来时可以更加简短

Snownee commented on src/main/java/snownee/lychee/util/recipe/LycheeRecipeSerializer.java:
```diff
@@ -0,0 +1,93 @@
+
+	RecordCodecBuilder<LycheeRecipe<?>, MinMaxBounds.Ints> MAX_REPEATS_CODEC =
+			MinMaxBounds.Ints.CODEC.fieldOf("max_repeats")
+								   .orElse(MinMaxBounds.Ints.ANY)
```
不同配方类型，范围应该是不同的

SettingDust replied:
这里和 https://github.com/Snownee/Lychee/blob/1.20-fabric/src/main/java/snownee/lychee/core/recipe/LycheeRecipe.java#L205-L208 是一个作用，作为默认值，之后会根据每个配方的实际序列化情况选择 Codec

Snownee replied:
> 之后会根据每个配方的实际序列化情况选择 Codec

应等待相关功能实现后再标记为“已解决”

Snownee commented on src/main/java/snownee/lychee/util/recipe/LycheeRecipeCommonHolder.java:
```diff
@@ -0,0 +1,16 @@
+public record LycheeRecipeCommonHolder(
+		boolean hideInRecipeViewer,
+		boolean ghost,
+		Optional<String> comment,
```
应该使用`@Nullable String`

SettingDust replied:
我是因为 record 会生成访问方法，没办法单独写个 getter 来返回 Optional，这里为什么需要 nullable string?

Snownee replied:
1. `optionalFieldOf`有直接提供默认值的方法
2. https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments

Snownee replied:
已修改代码

Snownee commented on src/main/java/snownee/lychee/util/contextual/ConditionHolder.java:
```diff
@@ -0,0 +1,52 @@
+									.forGetter(ConditionHolder::condition),
+							Codec.BOOL.fieldOf("secret")
+									  .orElse(false)
+									  .forGetter(ConditionHolder::secret),
```
把你的缩进改改吧，好丑

Snownee commented on src/main/java/snownee/lychee/util/contextual/ContextualConditionDisplay.java:
```diff
@@ -0,0 +1,69 @@
+import snownee.lychee.util.TriState;
+
+public interface ContextualConditionDisplay {
+	LoadingCache<ContextualConditionType<?>, String> DESCRIPTION_ID_CACHE =
```
此处缓存的必要较小，并且你没有考虑用户切换语言的情况

SettingDust replied:
嗯

Snownee commented on src/main/java/snownee/lychee/util/contextual/ContextualImpl.java:
```diff
@@ -0,0 +1,17 @@
+			RecordCodecBuilder.create(instance -> instance
+					.group(ContextualCommonHolder.CODEC
+							.fieldOf("conditions")
+							.orElse(new ContextualCommonHolder())
```
orElseGet

SettingDust committed:
chore: cleanup code

https://github.com/Snownee/Lychee/pull/70#discussion_r1477186819
https://github.com/Snownee/Lychee/pull/70#discussion_r1477202010

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
improve: nullable String in LycheeRecipeCommonHolder

Snownee committed:
feat: depends kiwi

SettingDust committed:
chore: remove BlockPredicate & use `optionalFieldOf`

https://github.com/Snownee/Lychee/pull/70#discussion_r1477340175
https://github.com/Snownee/Lychee/pull/70#discussion_r1477338699
Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
chore: delete docs

Snownee committed:
rename BlockInputLycheeRecipe -> BlockKeyableRecipe

Snownee committed:
prettier Codecs

Snownee committed:
`LycheeRecipe.allActions` now returns `Stream`

Snownee committed:
no more recipe id getter from LycheeRecipe

SettingDust committed:
refactor(context): allow context isn't have to be `LycheeContextValue`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor(recipe): convert `AnvilCraftingRecipe` to class for extendable

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee commented on src/main/java/snownee/lychee/recipes/AnvilCraftingRecipe.java:
```diff
@@ -190,4 +210,61 @@ private static AnvilCraftingRecipe of(
+				output,
+				assemblingActions
+		);
+	}
```
什么情况需要用到这个？

SettingDust committed:
refactor(recipe): using `applyCommonCodecs` instead of apply codecs manually

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore(recipe): cleanup code

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee commented on src/main/java/snownee/lychee/util/context/LycheeContext.java:
```diff
@@ -0,0 +1,68 @@
+		return Map.copyOf(context);
+	}
+
+	@Override
+	public int getContainerSize() {
+		return get(LycheeContextKey.ITEM).size();
+	}
+
+	@Override
+	public @NotNull ItemStack getItem(final int index) {
+		return get(LycheeContextKey.ITEM).get(index).get();
+	}
+
+	@Override
+	public void setItem(final int index, final ItemStack stack) {
+		get(LycheeContextKey.ITEM).replace(index, stack);
+	}
```
潜在的NPE

Snownee committed:
less interfaces, nicer codecs, more sanity

Snownee committed:
return of the good old ContextualHolder, and it uses less RAM

SettingDust reviewed (commented):
我觉得在这里考虑 RAM 有些过早优化了，我不认为这两种结构能造成多大的差异

Snownee commented:
> 我觉得在这里考虑 RAM 有些过早优化了，我不认为这两种结构能造成多大的差异

倒不如说为什么要把原来的写法改了。Lychee是一个有两年历史的项目。

SettingDust committed:
chore: action use common properties

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: remove generic for `ContextualCondition`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: using nullable description to avoid Optional field

https://github.com/Snownee/Lychee/pull/70#discussion_r1477338203
Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add `BlockCrushingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
improve: do not show prevent default description in tooltip if there are conditions in the action (close #71)

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
improve codecs

Snownee committed:
BlockPredicateCodec

SettingDust commented on src/main/java/snownee/lychee/util/codec/BlockPredicateCodec.java:
```diff
@@ -0,0 +1,60 @@
+	public static BlockPredicate fromString(String s) {
+		return CACHE.get(s, () -> {
+			if (s.equals("*")) {
+				//TODO how to handle ANY?
```
Codec 返回个 Optional？

Snownee replied:
等项目能运行起来之后再解决

Snownee committed:
TweakedItemStackResultCodec

Snownee committed:
TweakedItemStackCodec

Snownee committed:
randomly editing things

Snownee committed:
oops!

Snownee committed:
LocationPredicate

Snownee committed:
refactor LocationPredicate

Snownee committed:
no Optional

Snownee committed:
...

Snownee committed:
revive JSON fragments

Snownee committed:
delete unused mixins

Snownee committed:
use more shortcut method

SettingDust committed:
feat(recipe): add `BlockExplodingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(context): add required context type

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add `ShapedCraftingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix(recipe): adapt anvil mixin to new one(no test)

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore(recipe): adapt crafting & block crushing mixins to new one(no test)

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
style: include the editorconfig from https://github.com/Snownee/Loquat/blob/1.20.1-fabric/.editorconfig

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
style: reformat files

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add serializer and type to `ShapedCraftingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): use container generic instead of fixed `LycheeContext` for recipes & correct the `ShapedCraftingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add `BlockInteractingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add `BlockClickingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add `ItemBurningRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add `ItemExplodingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add `ItemInsideRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore(deps): bump minecraft and fabric to 24w07a

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: switch to new stream codec

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add serializer to `ItemExplodingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix(recipe): correct `BlockClickingRecipe` type and `invoke`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix(recipe): add `LightningChannelingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix(recipe): add `RandomBlockTickingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): add `DripstoneRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor(recipe): remove unused recipes

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee commented on src/main/java/snownee/lychee/util/predicates/BlockPredicateExtensions.java:
```diff
@@ -112,16 +113,18 @@ public static List<ItemStack> matchedItemStacks(BlockPredicate predicate) {
 	 * Optimized without get block state and block entity calls. And needn't pos loaded.
 	 */
-	public static boolean matches(BlockPredicate predicate, LycheeContext context) {
+	public static boolean matches(Level level, BlockPredicate predicate, LycheeContext context) {
```
为什么不从LycheeContext里获取？

SettingDust replied:
不知道 Recipe#matches 传入的 Level 和 context 里的是不是一致

Snownee commented on src/main/java/snownee/lychee/mixin/recipes/dripstone/BlockStateMixin.java:
```diff
@@ -0,0 +1,30 @@
+ * No DF lib for now 2024/2/28
+ */
+@Mixin(BlockStateBase.class)
+public class BlockStateMixin {
```
如果DF lib还更新的话，可以在MixinPlugin里选择性加载

SettingDust committed:
refactor(recipe): remove level arg for `BlockPredicateExtensions.matches`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix(recipe): missing level for `DripstoneRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: use Minecraft's predicates in 24w09a

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee commented on src/main/resources/lychee.mixins.json:
为什么缩进改了？

SettingDust committed:
style: correct indent

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: mark `blockexploding.ExplosionMixin` as need rewrite

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: move `LightningBoltMixin` into recipes

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): try to fix the `snownee.lychee.mixin.recipes.blockexploding.ExplosionMixin`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(recipe): don't touch the original drops

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `PreventDefault`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `SetItem`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `AddItemCooldown` & move conditions to `PostActionCommonProperties`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `AnvilDamageChance``

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix(recipe): adapt new process

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `Break`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `CustomAction`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): records to classes

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `PlaceBlock`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `CycleStateProperty`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(context): add recipe and id to context

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `Delay`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `DropItem`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): use record instead of classes

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `DropXp`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `Execute`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `Explode`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `Hurt`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `If`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `MoveTowardsFace`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `RandomSelect`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat(action): add `DamageItem`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: cleanup mixins

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: port AlwaysTrueIngredient

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: correct the recipe context

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: remove nbt patch stuffs

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: make `SCustomLevelEventPacket` work

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore(deps): bump dependencies

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: cleanup code make project runnable

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: update `SCustomLevelEventPacket` for new kiwi networking

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: update for new kiwi & 24w11a

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: revert nullable changes for using optional

Minecraft don't allow null as default value for optional field

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: init context key and serializer

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: CycleStateProperty$Type codec is null

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: wrong logic for required context construct

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: lack default value for time

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: wrong logic for required context construct

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: put recipe to context when possible

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: downgrade `onConstructed` to the implementation class

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: fix the ingredient pair codec

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: fix the single ingredient codec

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: make item stack codec can be plain string

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: the location contextual codec is wrong

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
revert: revert the namespace changes for actions and contextual

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: the result of anvil is wrong

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: redirect the hashcode and toString

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: new block predicate codec for supporting plain string

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: clearer description

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: switch to strict optional field

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: add offset codec

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: fix the `BlockInteractingRecipe#getIngredients`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: optimize the `BlockPredicate` codec for better debug info

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: correct the method to get matched stacks

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: changed the block in tag to blocks

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: make the block predicate codec with registry

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: make the block predicate codec with registry

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: use exact Y for fall distance

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: `ItemStackHolder#shrink` returned the wrong stack

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: `BlockCrushingRecipeType` showing wrong particle since the entity is AIR after matching

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: `BlockKeyableRecipeType` missing recipe context

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: `ExplosionMixin` missing level context

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: support flat item codec

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: explosion won't add drop when can't match recipe

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: `RandomSelect` is wrong

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: return ingredients correctly

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: `ItemInsideRecipe` optional block predicate

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: rewrite some dangerous condition

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: test of `DamageItem` is wrong

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: rename `ItemStackHolder#shrink` to `split`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add back getItemIndexes

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: allow 0 in lychee

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: apply post for `ShapedCraftingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: init the action data

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: add missing dripstone source

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: the conditions can't edit

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: send exploding packet & fix burning match

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: ItemShapeless dead loop

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add optional ingredient codec

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: rewrite the custom action/condition codec

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: swing when interacting

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: fix some builtin data pack entries

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: fix some builtin data pack entries

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
fix: info badge position in block exploding category

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: make the Hurt source optional

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: start works on REI compat

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: state codec need string

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: render the frames

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: avoid null & fix the title

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add display for block crushing

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: port ItemShapelessRecipeCategory

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
refactor: use interface for LycheeCategory

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add `LightningChannelingRecipe` category

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add `ItemExplodingRecipe` category

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: register workstation for `LightningChannelingRecipe`, `ItemExplodingRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add `BlockExplodingRecipe` category

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add `BlockInteractingRecipe` and `ItemBurningRecipe`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add `ItemInsideRecipeCategory`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add `DripstoneRecipeCategory`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: add item inside icon & fix block interaction ingredients

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: cleanup code

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: adapt to u/v scale in latest MC

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: avoid calling architectury methods

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: showing display name instead of key for difficulties

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: remove underscore in path

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: rewrite the display for list holder since the value mainly use in tooltips that can't hover

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: tweak the block lighting

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
improve: show the conditions for catalysts

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: disable recipes in lycheev

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: remove lycheev

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
fix

SettingDust committed:
fix: use max value & check if the min equals to max

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: use the correct display height

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: the info rect calc is wrong

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: add displays for anvil crafting and shaped crafting

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix: move `loadCatalystsInfo` to `PostActionRenderer` for avoiding load client class

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: avoid tag key in block predicate

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
fix: recipes

Snownee committed:
refactor: BlockPredicateExtensions string parsing

Snownee committed:
improve: cache Level in LycheeContext

Snownee committed:
fix: the `always_true` ingredient type does require testing

Snownee committed:
feat: VisualOnlyComponentsIngredient

Snownee committed:
improve: remove the random position offset when spawning item entity

Snownee committed:
improve: some codecs

Snownee committed:
feat: specify item to cool down in AddItemCooldown action

Snownee committed:
fix: data gen

Snownee committed:
refactor: re-implement recipe comparators

Snownee committed:
feat: `hide` now is a common property in actions

SettingDust committed:
fix: missing default renderer for actions

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
chore: small tweaks to the build script

Snownee commented:
moved the discussion to discord

SettingDust committed:
refactor: avoid optional for block predicate

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
fix(recipes): fix invalid `matches`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: remove `hashCode` and `equals`

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee commented on src/main/java/snownee/lychee/action/PlaceBlock.java:
```diff
@@ -154,12 +153,12 @@ public Component getDisplayName() {
 	@Override
 	public List<BlockPredicate> getOutputBlocks() {
-		return block.map(List::of).orElse(List.of());
+		return block == null ? List.of() : List.of(block);
```
block应该是非空的

Snownee commented on src/main/java/snownee/lychee/compat/rei/category/BlockCrushingRecipeCategory.java:
```diff
@@ -71,7 +70,7 @@ public List<Widget> setupDisplay(LycheeDisplay<BlockCrushingRecipe> display, Rec
 		widgets.add(Widgets.createDrawableWidget((GuiGraphics graphics, int mouseX, int mouseY, float delta) -> {
 			var x = recipe.getIngredients().isEmpty() ? 41 : 77;
-			var anyLandingBlock = recipe.landingBlock().isEmpty();
+			var anyLandingBlock = recipe.landingBlock() == BlockPredicateExtensions.ANY;
```
应该用BlockPredicateExtensions.isAny()

SettingDust committed:
refactor: use `BlockPredicateExtensions.isAny`

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: port to 1.20.5 pre1

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: update fabric api

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
feat: port to 1.20.5 pre3

Signed-off-by: SettingDust <settingdust@gmail.com>

Snownee committed:
improve: codecs, part 1

Snownee committed:
feat: 'contextual' -> 'if'

SettingDust committed:
refactor: remove compact list codec

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: fabric recipe api interface injection isn't working

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: bump minecraft to 1.20.5 pre4

Signed-off-by: SettingDust <settingdust@gmail.com>

SettingDust committed:
chore: disable REI runtime for running on 1.20.5

Notice, can't run since there isn't kiwi for 1.20.5 pre

Signed-off-by: SettingDust <settingdust@gmail.com>
